File: javascript/easy/closestBinarySearchTreeValue.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} target
 * @return {number}
 */
var closestValue = function(root, target) {
    return findClosest(root, target)[0].val;
};

var findClosest = function(root, target) {
    if(root == null) return [null, Number.MAX_SAFE_INTEGER];
    let delta = (root.val - target), nextNode = (delta > 0) ? root.left : root.right;
    let res = findClosest(nextNode, target);
    return (Math.abs(delta) < res[1]) ? [root, Math.abs(delta)] : res;
};

File: ruby/easy/closest_binary_search_tree_value.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {Float} target
# @return {Integer}
def closest_value(root, target)
    return 0 unless root
    min, min_root = (root.val - target).abs, root
    while(root)
        delta = root.val - target
        if(delta.abs < min)
            min = delta.abs
            min_root = root
        end
        root = (delta > 0) ? root.left : root.right
    end
    return min_root.val
end
